### ISSUE TITLE
Inheritance diagram extension class names should be resolved with __qualname__, not __name__

### ISSUE THREAD

**Issue by KM4YRI** [sphinx-doc/sphinx#5538]
### Problem

The inheritance diagram extension displays nested class names incorrectly, in the final diagram image.  Nested classes are only displayed as the final class name, instead of the fully qualified path.

Per [PEP 3155](https://www.python.org/dev/peps/pep-3155/), the `__qualname__` attribute is a superior way to resolve a string name of nested classes.

#### Procedure to reproduce the problem

1. Create a new python package called `foo_package`, add `__init__.py`, `doc` folder, etc..
2. Create a new sphinx project for `foo_package` under `/doc/`
2. In `conf.py`, set it up to point to the new package, and add `sphinx.ext.inheritance_diagram` to the `extensions` list.
3. Add the code below into `foo_package` and `index.rst`:

`__init__.py` within `foo_package`:
```
class Foo:
    class Bar:
        pass
    class Baz(Bar):
        pass

```

`index.rst` for `foo_package`:
```
.. inheritance-diagram::
    foo_package.Foo.Bar
    foo_package.Foo.Baz
```

#### Error logs / results

No error / exception / traceback, however the final diagram is incorrect, as the classes will be rendered with the names:

* `foo_package.Bar` instead of `foo_package.Foo.Bar`
* `foo_package.Baz` instead of `foo_package.Foo.Baz`

#### Expected results

Classes in the diagram should be rendered with the fully-qualified name, e.g. nested classes should be joined with '.', e.g. `foo_package.Foo.Baz` instead of `foo_package.Baz`.

### Environment info
- OS: Windows 7
- Python version: 3.5.1
- Sphinx version: 1.8.1 (with a fix applied that prevented the released version from running with 3.5.1)

### Proposed solution

Update the method `class_name()` within `inheritance_diagram.py` to use `__qualname__` instead of `__name__`.

Proposed change to `class_name()` below:

https://github.com/sphinx-doc/sphinx/blob/master/sphinx/ext/inheritance_diagram.py#L219

```
    def class_name(self, cls, parts=0, aliases=None):
        # type: (Any, int, Optional[Dict[unicode, unicode]]) -> unicode
        """Given a class object, return a fully-qualified name.

        This works for things I've tested in matplotlib so far, but may not be
        completely general.
        """
        module = cls.__module__
        if module in ('__builtin__', 'builtins'):
            fullname = cls.__name__
        else:
            fullname = '%s.%s' % (module, cls.__qualname__)
        if parts == 0:
            result = fullname
        else:
            name_parts = fullname.split('.')
            result = '.'.join(name_parts[-parts:])
        if aliases is not None and result in aliases:
            return aliases[result]
        return result
```

**Comment by bwohlberg** [comment]
The same problem applies to the ``show-inheritance`` option, which constructs incorrect base class names for nested classes, resulting in broken cross-references in the output.

As above, the simplest fix for Python 3 is to replace the line
```u':class:`%s.%s`' % (b.__module__, b.__name__)```
with
```u':class:`%s.%s`' % (b.__module__, b.__qualname__)```
in ``add_directive_header()`` in
[https://github.com/sphinx-doc/sphinx/blob/master/sphinx/ext/autodoc/__init__.py#L1115](https://github.com/sphinx-doc/sphinx/blob/master/sphinx/ext/autodoc/__init__.py#L1115).

While ``__qualname__`` is not available in  Python 2, it can be emulated by using the [qualname](https://github.com/wbolster/qualname) package.

**Comment by tk0miya** [comment]
+1; On next major release 2.0, we'll drop python2 support. So it would be nice to use qualname to its name!

Could you send us PR please?

**Comment by bwohlberg** [comment]
Done - see PR #5637.

**Comment by tk0miya** [comment]
This is already fixed by #5637 .
Thank you for your report!

**Comment by KM4YRI** [comment]
That's great news, thanks so much!
